### PR TITLE
Type annotate `torch.nn.Module` ctor

### LINKED ISSUE #45497: torch.Module typing __init__ no return value
__init__ return value should be specified as None. Now it doesn't have any
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L223

https://www.python.org/dev/peps/pep-0484/#the-meaning-of-annotations

cc @albanD @mruberry @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by malfet**
Annotate generic types
Fix some type violations
Override `_modules` and `_parameters` in `Sequential`, `ModuleList`, `ModuleDict`, etc

Dismantle pyramid of doom in `Module._apply`, by replacing:
```
for k,v in dict_.items():
   if v is not None:
      ....
```
with
```
for k,v in dict_.items():
   if v is None:
      continue
   ....
```
Fixes #45497

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -571,7 +571,8 @@ def compute_should_use_set_data(tensor, tensor_applied):
                         param.grad.data = grad_applied
                     else:
                         assert param.grad.is_leaf
-                        self._parameters[key].grad = grad_applied.requires_grad_(param.grad.requires_grad)
+                        self._parameters[key].grad = grad_applied.requires_grad_(  # type: ignore[union-attr]
```
Comment (albanD) [inline torch/nn/modules/module.py:574]: Why is this ignore? 
there is a `if param.grad is not None` just above so I guess it is not for the `.grad` field?

File: torch/nn/utils/spectral_norm.py [hunk 2]
```diff
@@ -120,7 +120,7 @@ def apply(module: Module, name: str, n_power_iterations: int, dim: int, eps: flo
 
         fn = SpectralNorm(name, n_power_iterations, dim, eps)
         weight = module._parameters[name]
-        if isinstance(weight, torch.nn.parameter.UninitializedParameter):
+        if isinstance(weight, torch.nn.parameter.UninitializedParameter) or weight is None:
```
Comment (albanD) [inline torch/nn/utils/spectral_norm.py:123]: That's not really the right error when `weight is None`.
For this one, we only want to say that you cannot apply spectral_norm on a Parameter that is None.

Good catch though!

Sadly, I note the inconsistency with the weight_norm that implements this slightly differently and so doesn't have type issue: https://github.com/pytorch/pytorch/blob/aa62d305de107f46e4148ed46c52f44a73212299/torch/nn/utils/weight_norm.py#L38
(you an ignore it for this PR).

**Review by albanD — COMMENTED** [review]
Looks good.
Just an error message update.

File: torch/nn/modules/module.py (hunk 1 above)
Comment (malfet) [reply torch/nn/modules/module.py:574]: python/mypy#10784

Comment (malfet) [reply torch/nn/modules/module.py:574]: python/mypy#10784

Comment (albanD) [reply torch/nn/modules/module.py:574]: Ho very good point from them.
I guess we should replace the `self._parameters[key]` with `param` here then?

Comment (malfet) [reply torch/nn/modules/module.py:574]: Actually as `.items()` gives reference to an object, this is a very good point, updated.

File: torch/nn/modules/container.py [hunk 3]
```diff
@@ -419,6 +425,8 @@ def forward(self, x):
                 return x
     """
 
+    _parameters: Dict[str, 'Parameter']  # type: ignore[assignment]
```
Comment (albanD) [inline torch/nn/modules/container.py:428]: This should already be on the parents class "Module" no?

File: torch/nn/modules/container.py [hunk 4]
```diff
@@ -561,6 +569,8 @@ def forward(self, x, choice):
...
```
Comment (albanD) [inline torch/nn/modules/container.py:572]: Same as above

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -247,23 +247,23 @@ def forward(self, x):
...
         self.training = True
-        self._parameters = OrderedDict()
-        self._buffers = OrderedDict()
-        self._non_persistent_buffers_set = set()
-        self._backward_hooks = OrderedDict()
+        self._parameters: Dict[str, Optional[Parameter]] = OrderedDict()
+        self._buffers: Dict[str, Optional[Tensor]] = OrderedDict()
+        self._non_persistent_buffers_set: Set[str] = set()
+        self._backward_hooks: Dict[int, Callable] = OrderedDict()
         self._is_full_backward_hook = None
```
Comment (albanD) [inline torch/nn/modules/module.py:261]: This one should be an Optional[bool] I think

File: torch/nn/modules/container.py (hunk 3 above)
Comment (malfet) [reply torch/nn/modules/container.py:428]: No, on 'Module',  'Parameter' is optional, while in ParameterList it is mandatory

File: torch/nn/modules/module.py (hunk 5 above)
Comment (malfet) [reply torch/nn/modules/module.py:261]: It is (and you've annotated it ;)), see 
https://github.com/pytorch/pytorch/blob/287603f51cd0303f267f1f978bcab71aa07e3d77/torch/nn/modules/module.py#L248

**Review by albanD — APPROVED** [review]
LGTM thanks for the update

**Pull request closed without merging**